### PR TITLE
38 Create the Quiz Set UI

### LINKED ISSUE #38: Create the Quiz Set UI
As a user, I want to choose a category of questions to quiz, in order to focus on specific questions.

**Acceptance Criteria:**
- [x] The existing start quiz screen is replaced with this new screen
- [x] All categories are listed + a shuffle and exam mode
- [x] When clicking a category a popup shows up
  - [x] A timer can be enabled/disabled
    - [x] Time can be set or chosen by the user
  - [x] The number of questions can be chosen (short, normal, all)
  - [x] The quiz can be started (like on the old screen but now with the options)
  - [x] For the exam quiz these options are disabled and the chosen options are indicated
- [x] The latest quiz scores of each category is shown if available

### PR DESCRIPTION
**Pull request by plepisnew**
Closes #38

- Added options for taking an exam, quiz, and reviewing questions
- Added timer and question number selection for non-exam options
- Added all categories for taking a quiz

### :ballot_box_with_check: Definition of Done checklist
- [x] Meaningful title and valuable description in pull request
- [x] Code change tested manually if not covered by automated tests
- [x] Pull request linked to one or more issues/stories
- [x] Acceptance criteria of the linked stories satisfied
- [x] Code change comprehended by those who approved it
- [x] State of target branch maintained or improved after the code change

### CODE REVIEW

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java [hunk 1]
```diff
@@ -125,32 +131,15 @@ protected void onCreate(Bundle savedInstanceState) {
         bottomNavigationView.setSelectedItemId(R.id.practiceTheory);
 
         //add questions to question list via helper method --> help us select question
-        db = new CrushersDataBase(this);
+        CrushersDataBase db = new CrushersDataBase(this);
         questionList = db.getAllQuestions();
-        Collections.shuffle(questionList);
```
Comment (GeorgZs) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:130]: Why was this removed, the intention was to spice up the questions when a user restarts a quiz; otherwise they will simply just keep going through the same sequence of questions

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java [hunk 2]
```diff
@@ -125,32 +131,15 @@ protected void onCreate(Bundle savedInstanceState) {
...
-            if(!(categorySelected.equals("No Category")))
-                customQuiz(questionCountSelected, categorySelected);
-            else customQuiz(questionCountSelected, "No Category");
-        else {
-            //else start a random quiz -- No settings selected
-            quizState = new QuizState(true, questionList, null, scoreNumber);
-            totalQuestions = questionList.size();
-            addQuestion(questionList);
-        }
+        customQuiz(numberOfQuestions, category);
```
Comment (GeorgZs) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:144]: How are you handling quizes then where a person does not select a category or/and select a number of questions then?

Comment (GeorgZs) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:144]: Since if you solely pass them through without handling them they will create an exception since expecially for numberOfQuestions it will pass a 0 if there is nothing passed through and then the custom quiz fails on getting 0 questions

**Review by GeorgZs — CHANGES_REQUESTED** [review]
Please read the comments regarding the custom quiz handling

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java (hunk 1 above)
Comment (GeorgZs) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:130]: Everyday im shuffling XD

Comment (plepisnew) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:130]: > Why was this removed, the intention was to spice up the questions when a user restarts a quiz; otherwise they will simply just keep going through the same sequence of questions

Good catch. It was throwing NPE and I must've accidentally deleted it instead of commenting out

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java (hunk 2 above)
Comment (plepisnew) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:144]: > How are you handling quizes then where a person does not select a category or/and select a number of questions 

image

`PracticeTheoryActivity::startQuiz` When the user starts a Quiz, if a category hasn't been selected, it will default to "No Category" (which is then passed to the QuizActivity using intents). Is there a better way to do this? Yes. Yes, there is xD

Comment (GeorgZs) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/QuizQuestionActivity.java:144]: Perfect yea once these are fixed thats all good then my g

**Comment by plepisnew** [comment]
> Please read the comments regarding the custom quiz handling

👍

**Review by GeorgZs — APPROVED** [review]
noice
